### PR TITLE
refactor: redo CSX merge logic layout and names

### PR DESCRIPTION
**Pull request by WalterBright**
The idea is to:

* push CSX logic into ctorflow.d
* push error handling into dscope.d
* use `a` and `b` for left and right branches

### CODE REVIEW

File: src/dmd/ctorflow.d [hunk 1]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
- *      fieldInit = the path to merge fi into
- *      fi = the other path
+ *      a = the path to merge `b` into
+ *      b = the other path
  * Returns:
- *      false means either fieldInit or fi skips initialization
+ *      false means one of the paths skips construction
  */
-bool mergeFieldInitX(ref CSX fieldInit, const CSX fi)
+bool mergeCallSuper(ref CSX a, const CSX b) pure nothrow
```
Comment (jacob-carlborg) [inline src/dmd/ctorflow.d]: Why not call it `leftBranch` and `rightBranch`, if that is what it means?

Comment (wilzbach) [reply src/dmd/ctorflow.d]: Maybe `lhs` and `rhs` if brevity is really needed?

Comment (WalterBright) [reply src/dmd/ctorflow.d]: Please, no.

* What follows in the implementation is essentially math, and in math one uses one letter variable names for good reason. We write `sin x`, not `sin theAngleInRadians`.

* One letter names are easiest to comprehend when one has complex math expressions.

* The algorithm is commutative, i.e. there is no `left` or `right`. Giving such names implies there is an asymmetry when there is none.

* Also, it's good style to use shorter names the more local they are.

File: src/dmd/ctorflow.d [hunk 2]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
         {
-            ok = true;
-            fieldInit = CSX.halt;
+            a = CSX.halt;
         }
-        else if (!aHalt && aRet)
+        else if ((!bHalt && bRet && !bAny && aAny) || (!aHalt && aRet && !aAny && bAny))
+        {
+            // If one has returned without a constructor call, there must be never
+            // have been ctor calls in the other.
```
Comment (JinShil) [inline src/dmd/ctorflow.d]: "there must be never have been" doesn't make sense to me.

File: src/dmd/ctorflow.d (hunk 1 above)
Comment (andralex) [reply src/dmd/ctorflow.d]: The names are fine. The code is wanting. Reviewers should suggest material improvement, not trivial pursuits.

File: src/dmd/ctorflow.d [hunk 3]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
-bool mergeFieldInitX(ref CSX fieldInit, const CSX fi)
+bool mergeCallSuper(ref CSX a, const CSX b) pure nothrow
 {
-    if (fi != fieldInit)
+    // This does a primitive flow analysis to support the restrictions
+    // regarding when and how constructors can appear.
+    // It merges the results of two paths.
+    // The two paths are `a` and `b`; the result is merged into `a`.
+    bool ok = true;
+    if (b != a)
```
Comment (andralex) [inline src/dmd/ctorflow.d]: The function does nothing and returns `true` if `a == b`, yet it takes the reader a whopping 47 lines to figure that out. Can we please please please with Truvia on top renounce this crap:

```d
bool ok = true;
if (condition)
{
    ... so many lines ...
}
return ok;
```

This antipattern is all over the compiler code. It's insufferable in old code, and definitely not tolerable in the sligthest in new code. We're not coding Fortran over here. Please:

```d
if (!condition)
{
    return true;
}
... knock yourself out ...
```

File: src/dmd/ctorflow.d [hunk 4]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
     {
+        // Have ALL branches called a constructor?
+        int aAll = (a & (CSX.this_ctor | CSX.super_ctor)) != 0;
+        int bAll = (b & (CSX.this_ctor | CSX.super_ctor)) != 0;
```
Comment (andralex) [inline src/dmd/ctorflow.d]: These are `bool` really, not `int`. And they should be immutable. Just use `immutable` or `const` and let type deduction say what you mean.

File: src/dmd/ctorflow.d [hunk 5]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
+        // Have ANY branches called a constructor?
+        bool aAny = (a & CSX.any_ctor) != 0;
+        bool bAny = (b & CSX.any_ctor) != 0;
```
Comment (andralex) [inline src/dmd/ctorflow.d]: immutable throughout

File: src/dmd/ctorflow.d [hunk 6]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
-            ok = true;
-            fieldInit = CSX.halt;
+            a = CSX.halt;
         }
-        else if (!aHalt && aRet)
+        else if ((!bHalt && bRet && !bAny && aAny) || (!aHalt && aRet && !aAny && bAny))
+        {
+            // If one has returned without a constructor call, there must not
+            // be ctor calls in the other.
+            ok = false;
```
Comment (andralex) [inline src/dmd/ctorflow.d]: Just return - all of this `ok` variable nonsense is another antipattern

File: src/dmd/ctorflow.d [hunk 7]
```diff
@@ -105,53 +105,118 @@ struct CtorFlow
...
+ * Merge `b` flow analysis results into `a`.
+ * Params:
+ *      a = the path to merge `b` into
+ *      b = the other path
+ * Returns:
+ *      false means either `a` or `b` skips initialization
+ */
+bool mergeFieldInit(ref CSX a, const CSX b) pure nothrow
+{
+    if (b != a)
```
Comment (andralex) [inline src/dmd/ctorflow.d]: here we go again

**Review by andralex — CHANGES_REQUESTED** [review]

File: src/dmd/ctorflow.d (hunk 7 above)
Comment (WalterBright) [reply src/dmd/ctorflow.d]: Well, I beg that that code was all copy-pasta from the original.

**Review by RazvanN7 — APPROVED** [review]

**Review by andralex — APPROVED** [review]
Nice, thx! So what's with the failures?

**Comment by RazvanN7** [comment]
@andralex the circleCi failure seems unrelated. @WalterBright please give it another push so that you will restart the failing test.

**Comment by wilzbach** [comment]
You can restart CircleCi directly by going there, logging in via GitHub and hitting restart.

**Comment by wilzbach** [comment]
(I just did so)

**Pull request merged**